x-pack/plugin/snapshot-repo-test-kit/src/main/java/org/elasticsearch/repositories/blobstore/testkit/RepositoryAnalyzeAction.java

```diff
@@ -0,0 +1,1013 @@
+                        + request.getReroutedFrom()
+                        + "] to non-snapshot node"
+                );
+            }
+
+            request.reroutedFrom(localNode);
+            final List<DiscoveryNode> snapshotNodes = getSnapshotNodes(state.nodes());
+            if (snapshotNodes.isEmpty()) {
+                listener.onFailure(
+                    new IllegalArgumentException("no snapshot nodes found for analysis of repository [" + request.getRepositoryName() + "]")
+                );
+            } else {
+                if (snapshotNodes.size() > 1) {
+                    snapshotNodes.remove(state.nodes().getMasterNode());
+                }
+                final DiscoveryNode targetNode = snapshotNodes.get(new Random(request.getSeed()).nextInt(snapshotNodes.size()));
+                RepositoryAnalyzeAction.logger.trace("rerouting analysis [{}] to [{}]", request.getDescription(), targetNode);
+                transportService.sendChildRequest(
+                    targetNode,
+                    NAME,
+                    request,
+                    task,
+                    TransportRequestOptions.EMPTY,
+                    new ActionListenerResponseHandler<>(listener, Response::new)
+                );
+            }
+        }
+    }
+
+    private static boolean isSnapshotNode(DiscoveryNode discoveryNode) {
+        return (discoveryNode.isDataNode() || discoveryNode.isMasterNode())
+            && RepositoriesService.isDedicatedVotingOnlyNode(discoveryNode.getRoles()) == false;
+    }
+
+    private static List<DiscoveryNode> getSnapshotNodes(DiscoveryNodes discoveryNodes) {
+        final ObjectContainer<DiscoveryNode> nodesContainer = discoveryNodes.getMasterAndDataNodes().values();
+        final List<DiscoveryNode> nodes = new ArrayList<>(nodesContainer.size());
+        for (ObjectCursor<DiscoveryNode> cursor : nodesContainer) {
+            if (isSnapshotNode(cursor.value)) {
+                nodes.add(cursor.value);
+            }
+        }
+        return nodes;
+    }
+
+    // Exposed for tests
+    static List<Long> getBlobSizes(Request request) {
+
+        int blobCount = request.getBlobCount();
+        long maxTotalBytes = request.getMaxTotalDataSize().getBytes();
+        long maxBlobSize = request.getMaxBlobSize().getBytes();
+
+        if (maxTotalBytes - maxBlobSize < blobCount - 1) {
+            throw new IllegalArgumentException(
+                "cannot satisfy max total bytes ["
+                    + maxTotalBytes
+                    + "B/"
+                    + request.getMaxTotalDataSize()
+                    + "]: must write at least one byte per blob and at least one max-sized blob which is ["
+                    + (blobCount + maxBlobSize - 1)
+                    + "B] in total"
+            );
+        }
+
+        final List<Long> blobSizes = new ArrayList<>();
+        for (long s = 1; 0 < s && s < maxBlobSize; s <<= 1) {
```

Comment:
Is there a reason why we are only using blobs with a power of 2 size?

Reply:
Not especially, it's just a convenient way to make blobs that are both large and small. It's possible this would miss some bug that only happens when the blob is not a power of two in size, for instance with multi-part uploads we won't often be testing cases where the final part is incomplete like this. I won't address this today but added a TODO in b3c6beeea5bd952a58d287d9af70384f438011f4.

Reply:
Yes, I was thinking in scenarios such as the one you've described. Let's tackle this in a follow-up PR. 👍